src/debouncer.rs

```diff
@@ -0,0 +1,198 @@
+}
+
+/// A debounced event.
+///
+/// Does not emit any specific event type on purpose, only distinguishes between an any event and a continuous any event.
+#[derive(Clone, Debug, Eq, Hash, PartialEq)]
+#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
+pub struct DebouncedEvent {
+    /// Event path
+    pub path: PathBuf,
+    /// Event kind
+    pub kind: DebouncedEventKind,
+}
+
+impl DebouncedEvent {
+    fn new(path: PathBuf, kind: DebouncedEventKind) -> Self {
+        Self { path, kind }
+    }
+}
+
+type DebounceData = Arc<Mutex<DebounceDataInner>>;
+
+#[derive(Default)]
+struct DebounceDataInner {
+    d: HashMap<PathBuf, EventData>,
+    timeout: Duration,
+    e: Vec<crate::Error>,
+}
+
+impl DebounceDataInner {
+    /// Retrieve a vec of debounced events, removing them if not continuous
+    pub fn debounced_events(&mut self) -> Vec<DebouncedEvent> {
+        let mut events_expired = Vec::with_capacity(self.d.len());
+        let mut data_back = HashMap::with_capacity(self.d.len());
+        // TODO: perfect fit for drain_filter https://github.com/rust-lang/rust/issues/59618
+        for (k, v) in self.d.drain() {
+            if v.update.elapsed() >= self.timeout {
+                println!("normal timeout");
+                events_expired.push(DebouncedEvent::new(k, DebouncedEventKind::Any));
+            } else if v.insert.elapsed() >= self.timeout {
+                println!("continuous");
+                data_back.insert(k.clone(), v);
+                events_expired.push(DebouncedEvent::new(k, DebouncedEventKind::AnyContinuous));
+            } else {
+                data_back.insert(k, v);
+            }
+        }
+        self.d = data_back;
+        events_expired
+    }
+
+    /// Returns all currently stored errors
+    pub fn errors(&mut self) -> Vec<Error> {
+        let mut v = Vec::new();
+        std::mem::swap(&mut v, &mut self.e);
+        v
+    }
+
+    /// Add an error entry to re-send later on
+    pub fn add_error(&mut self, e: crate::Error) {
+        self.e.push(e);
+    }
+
+    /// Add new event to debouncer cache
+    pub fn add_event(&mut self, e: Event) {
+        for path in e.paths.into_iter() {
+            if let Some(v) = self.d.get_mut(&path) {
+                v.update = Instant::now();
+                println!("Exists");
+            } else {
+                self.d.insert(path, EventData::new_any());
+            }
+        }
+    }
+}
+
+/// Creates a new debounced watcher.
+/// 
+/// Timeout is the amount of time after which a debounced event is emitted or a Continuous event is send, if there still are events incoming for the specific path.
+/// 
+/// If tick_rate is None, notify will select a tick rate that is less than the provided timeout.
+pub fn new_debouncer(
+    timeout: Duration,
+    tick_rate: Option<Duration>,
+) -> Result<(Receiver<DebounceChannelType>, RecommendedWatcher), Error> {
+    let data = DebounceData::default();
+
+    let tick_div = 4;
```

Comment:
nit: I'd like to manage this value as const and declare it on the module scope.

Reply:
understandable, had a moment of wrangling with myself whether I'd want that as top module declaration